pyramid/router.py

```diff
@@ -164,14 +164,38 @@ def handle_request(self, request):
             except PredicateMismatch:
                 # look for other views that meet the predicate
                 # criteria
-                for iface in context_iface.__sro__[1:]:
+                pred_matches = []
+
+                for index, iface in enumerate(context_iface.__sro__[1:]):
                     previous_view_callable = view_callable
                     view_callable = adapters.lookup(
                         (IViewClassifier, request.request_iface, iface),
                         IView, name=view_name, default=None)
+
                     # intermediate bases may lookup same view_callable
-                    if view_callable is previous_view_callable:
+                    if view_callable is previous_view_callable or \
+                       view_callable is None:
                         continue
+
+                    # if we have predicates, lets see how many match and try
+                    # most specific first.
+                    if hasattr(view_callable, '__predicates__'):
+                        count = 0
+                        for pred in view_callable.__predicates__:
+                            if pred(context, request):
+                                count += 1
```

Comment:
@tseaver I just kept it in because the code was already handling predicatemismatch down below but we can totally remove it, I'll push a separate commit